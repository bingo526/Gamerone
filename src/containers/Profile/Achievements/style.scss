$achievement-action-post-color: #5b3df2;
$achievement-action-gear-color: #4cd8b5;
$achievement-action-follower-color: #f24981;
$achievement-action-experience-color: #4cd8b5;
