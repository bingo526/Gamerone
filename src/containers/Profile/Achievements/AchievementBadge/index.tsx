import React from 'react';
import cn from 'classnames';
import Icon, { IconSize } from 'components/common/Icon';
import { ACHIEVEMENT_ICON_LIST } from 'utils/constants';
import { ActionType } from 'interfaces';
import Badge, { BadgeSize } from 'components/common/Badge';
import './style.scss';

export interface AchievementIconProps {
  action: ActionType;
  iconSize?: IconSize;
  badgeSize?: BadgeSize;
}

const AchievementBadge: React.FC<AchievementIconProps> = ({
  action,
  iconSize = '1x',
  badgeSize,
}: AchievementIconProps): JSX.Element => {
  const badgeClasses = cn({
    [`achievement-badge--${action}`]: action,
  });

  return (
    <div className={badgeClasses}>
      <Badge type="outline" size={badgeSize} testid="badge-achievement">
        <Icon name={ACHIEVEMENT_ICON_LIST[action]} size={iconSize} />
      </Badge>
    </div>
  );
};

export default React.memo(AchievementBadge);
