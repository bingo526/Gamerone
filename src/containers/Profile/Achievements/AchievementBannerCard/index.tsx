import React from 'react';
import { Achievement, ActionType } from 'interfaces';
import Card, { CardTypeEnum } from 'components/common/Card';
import Badge from 'components/common/Badge';
import './style.scss';
import { formatDateDayMonthYear } from 'utils/dateFormatter';

export interface AchievementBannerCardProps {
  action: ActionType;
  achievements: Achievement[] | null;
}

export const sortAchievementThreshold = (a: Achievement, b: Achievement) =>
  (b as Achievement).threshold - (a as Achievement).date.getTime();

const AchievementBannerCard: React.FC<AchievementBannerCardProps> = ({
  action,
  achievements,
}: AchievementBannerCardProps): JSX.Element => {
  const actionsOnly =
    achievements &&
    achievements
      .filter((a: Achievement) => a.action === action)
      .sort(sortAchievementThreshold);

  const achievement =
    actionsOnly && actionsOnly.length > 0 ? actionsOnly[0] : undefined;

  return achievement ? (
    <Card type={CardTypeEnum.BANNER}>
      <div
        className="card__content"
        style={{
          textAlign: 'center',
          paddingTop: '5rem',
          paddingBottom: '5rem',
        }}
      >
        <Badge size="xlarge" type="outline" testid="badge-banner">
          {achievement.action === 'follower' ? (
            <i className="icon icon-2x icon-video-game-controller-team" />
          ) : achievement.action === 'post' ? (
            <i className="icon icon-2x icon-quill-write" />
          ) : (
            ''
          )}
        </Badge>

        <br />
        <br />
        <br />

        <h3 className="number">{achievement.threshold}</h3>
        <h4 className="label">
          {achievement.action === 'follower'
            ? 'Followers'
            : achievement.action === 'post'
            ? 'Posts'
            : ''}
        </h4>

        <br />

        <time className="time">
          {formatDateDayMonthYear(new Date(achievement.date))}
        </time>
      </div>
    </Card>
  ) : (
    <Card type={CardTypeEnum.BANNER}>
      <div
        className="card__content"
        style={{
          textAlign: 'center',
        }}
      >
        <h4>No major {action} milestones yet.</h4>
      </div>
    </Card>
  );
};

export default AchievementBannerCard;
