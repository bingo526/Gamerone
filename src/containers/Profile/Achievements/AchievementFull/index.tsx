import React, { ReactNode } from 'react';
import Card, { CardTypeEnum } from 'components/common/Card';
import './style.scss';

export interface AchievementFullCardProps {
  children?: ReactNode;
}

const AchievementFullCard: React.FC<AchievementFullCardProps> = ({
  children,
}: AchievementFullCardProps): JSX.Element => {
  return (
    <Card type={CardTypeEnum.ACHIEVEMENTS_FULL}>
      <div className="card__header">
        <h4>All badges</h4>
      </div>
      <div className="card__content">{children}</div>
    </Card>
  );
};

export default AchievementFullCard;
