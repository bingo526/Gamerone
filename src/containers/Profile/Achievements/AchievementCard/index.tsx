import React from 'react';
import cn from 'classnames';
import { formatDateDayMonthYear } from 'utils/dateFormatter';
import { AchievementModel } from 'models/AchievementModel';
import AchievementBadge from '../AchievementBadge';
import './style.scss';

export interface AchievementCardProps {
  achievement: AchievementModel;
}

export const AchievementCard: React.FC<AchievementCardProps> = ({
  achievement,
}: AchievementCardProps): JSX.Element => {
  const badgeNameClasses = cn('badge-name', {
    [`badge-name--${achievement.action}`]: achievement.action,
  });

  return (
    <div className="achievement-card-wrapper">
      <AchievementBadge
        action={achievement.action}
        badgeSize="xlarge"
        iconSize="3x"
      />
      {achievement.badge && (
        <div className="achievement-details">
          <p className={badgeNameClasses}>{achievement.badge.name}</p>
          <p className="badge-date">
            {formatDateDayMonthYear(achievement.date)}
          </p>
        </div>
      )}
    </div>
  );
};

export default AchievementCard;
