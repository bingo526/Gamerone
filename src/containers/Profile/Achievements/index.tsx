import React from 'react';
import { connect } from 'react-redux';
import { selectCurrentProfileUser } from 'redux/profile/selectors';
import { RootState } from 'redux/types';
import { Achievement } from 'interfaces';
import usePromise from 'lib/usePromise';
import { getProfileAchievementsById } from 'api/achievements';
import { UserModel } from 'models/user';
import Loader from 'components/common/Loader';
import AchievementFullCard from './AchievementFull';
import AchievementCard from './AchievementCard';
import { AchievementModel } from 'models/AchievementModel';
import AchievementBannerCard from './AchievementBannerCard';
import './style.scss';
import OpenGraph, { PageMeta } from 'components/common/OpenGraph';
import { SITE_URL } from 'utils/constants';
import sort from 'fast-sort';

const Achievements: React.FC<MappedProps> = ({
  user,
}: MappedProps): JSX.Element => {
  const [isLoading, achievements, error] = usePromise<AchievementModel[]>(
    () =>
      new Promise(async (resolve, reject) => {
        try {
          const data = (await getProfileAchievementsById(
            user.id,
          )) as Achievement[];
          const achivements = data.map((item) =>
            new AchievementModel().fromDto(item),
          );

          sort(achivements).desc((ach) => ach.date.getTime());

          resolve(achivements);
        } catch (ex) {
          reject(ex);
        }
      }),
    [user],
  );

  const pageTitle = user?.username + "'s Achievements";
  const achievementMeta = {
    title:
      'Check out ' +
      user?.username +
      "'s " +
      achievements?.length +
      ' achievements.',
    description: user?.bio,
    image: user?.avatar,
    url: SITE_URL + '/' + user?.username + '/achievements',
    type: 'profile',
    username: user?.username,
    firstName: user?.firstName,
    lastName: user?.lastName,
  } as PageMeta;

  return (
    <>
      <OpenGraph title={pageTitle} meta={achievementMeta} />
      <AchievementBannerCard action="follower" achievements={achievements} />
      <AchievementFullCard>
        {isLoading && !error && <Loader />}
        {!isLoading && achievements?.length === 0 && <h3>No Achievements</h3>}
        {achievements != null &&
          achievements.map((achievement) => (
            <AchievementCard key={achievement.id} achievement={achievement} />
          ))}
      </AchievementFullCard>
      <AchievementBannerCard action="post" achievements={achievements} />
    </>
  );
};

const mapStateToProps = (state: RootState) => {
  return {
    user: selectCurrentProfileUser(state) || new UserModel(),
  };
};

type MappedProps = ReturnType<typeof mapStateToProps>;

export default connect(mapStateToProps)(Achievements);
