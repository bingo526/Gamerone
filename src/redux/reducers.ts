import Auth from './auth/reducer';
import Profile from './profile/reducer';
import Settings from './settings/reducer';
import Post from './post/reducer';
import RequestStatus from './request-status/reducer';
import PostForm from './post-form/reducer';
import Dialog from './dialogs/reducer';
import Misc from './misc/reducer';

export default {
  Auth,
  Profile,
  Settings,
  Post,
  PostForm,
  Dialog,
  RequestStatus,
  Misc,
};
