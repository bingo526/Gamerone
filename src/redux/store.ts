import { createStore, combineReducers, applyMiddleware, compose } from 'redux';
import { createBrowserHistory as createHistory } from 'history';
import { connectRouter, routerMiddleware } from 'connected-react-router';
import thunk from 'redux-thunk';
import createSagaMiddleware from 'redux-saga';
import reducers from './reducers';
import rootSaga from './sagas';

const history = createHistory();
const sagaMiddleware = createSagaMiddleware();
const routeMiddleware = routerMiddleware(history);
const middlewares = [thunk, sagaMiddleware, routeMiddleware];

const composeEnhancers =
  (window['__REDUX_DEVTOOLS_EXTENSION_COMPOSE__'] as typeof compose) || compose;

const store = createStore(
  combineReducers({
    ...reducers,
    router: connectRouter(history),
  }),
  composeEnhancers(applyMiddleware(...middlewares)),
);
sagaMiddleware.run(rootSaga);
export { store, history };
