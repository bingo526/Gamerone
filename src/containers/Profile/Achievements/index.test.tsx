import Achievements from './index';

import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { render, fireEvent } from '@testing-library/react';
import { act } from 'react-dom/test-utils';

import { store } from 'redux/store';

const templateAchievementsData = [
  {
    id: 1,
    date: null,
    badge: {
      id: 1,
      name: 'bir',
      description: 'cds',
      logo: 'badges/placeholder.svg',
    },
  },
  {
    id: 3,
    date: null,
    badge: {
      id: 2,
      name: 'iki',
      description: 'fgergfwegwergewrgwe',
      logo: 'badges/placeholder.svg',
    },
  },
];

const renderForm = async () => {
  let utils;

  await act(async () => {
    utils = render(
      <Provider store={store}>
        <MemoryRouter>
          <Achievements />
        </MemoryRouter>
      </Provider>,
    );
  });

  // eslint-disable-next-line @typescript-eslint/ban-ts-ignore
  // @ts-ignore
  return { ...utils };
};

test('Fetch achievements data from API', async () => {
  const spy = jest
    .spyOn(window, 'fetch')
    .mockImplementationOnce(async (url) => {
      return Promise.resolve({
        status: 200,
        json: () => templateAchievementsData,
      });
    });

  const utils = await renderForm();

  expect(utils.asFragment()).toMatchSnapshot();

  await spy.mockRestore();
});
