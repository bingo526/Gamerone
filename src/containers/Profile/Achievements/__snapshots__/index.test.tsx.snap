// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Fetch achievements data from API 1`] = `
<DocumentFragment>
  <div
    class="grid grid__profile"
  >
    <div
      class="grid__spacer-2"
    />
    <div
      class="card card--achievements"
    >
      <div
        class="card__content"
      >
        <img
          alt="bir"
          height="128"
          src="CDN_ENV_NOT_DEFINED/128x128badges/placeholder.svg"
          title="bir"
          width="128"
        />
        bircds
      </div>
    </div>
    <div
      class="card card--achievements"
    >
      <div
        class="card__content"
      >
        <img
          alt="iki"
          height="128"
          src="CDN_ENV_NOT_DEFINED/128x128badges/placeholder.svg"
          title="iki"
          width="128"
        />
        ikifgergfwegwergewrgwe
      </div>
    </div>
  </div>
</DocumentFragment>
`;
